bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/Connection.java

```diff
@@ -1553,87 +1558,97 @@ public void executeSequenceCommand(@Nullable Device device, String command,
-            if (singleGroupTimer != null) {
-                singleGroupTimer.cancel(true);
-                singleGroupTimer = null;
+        QueueObject queueObject = new QueueObject();
+        queueObject.devices = devices;
+        queueObject.nodeToExecute = nodeToExecute;
+        String serial = "";
+        for (Device device : devices) {
+            if (!this.devices.containsKey(device.serialNumber)) {
+                this.devices.put(device.serialNumber, new LinkedBlockingQueue<>());
             }
-            singleGroupTimer = scheduler.schedule(() -> executeSequenceNode(devices, nodeToExecute), 500,
-                    TimeUnit.MILLISECONDS);
-
-            return;
+            this.devices.get(device.serialNumber).offer(queueObject);
+            serial = serial + device.serialNumber + " ";
         }
+        logger.debug("added {} device {}", queueObject.hashCode(), serial);
+    }
 
-        if (devices.length == 1) {
-            if (!singles.containsKey(devices[0])) {
-                singles.put(devices[0], new QueueObject());
-            }
-            singles.get(devices[0]).queue.add(nodeToExecute);
-        } else {
-            if (!groups.containsKey(devices[0])) {
-                groups.put(devices[0], new QueueObject());
+    private void handleExecuteSequenceNode() {
+        if (devicesRunning.compareAndSet(false, true)) {
+            try {
+                for (String serialNumber : devices.keySet()) {
+                    LinkedBlockingQueue<QueueObject> queueObjects = devices.get(serialNumber);
+                    QueueObject queueObject = queueObjects.peek();
+                    if (queueObject != null && (queueObject.future == null || queueObject.future.isDone())) {
+                        boolean execute = true;
+                        String serial = "";
+                        for (Device tmpDevice : queueObject.devices) {
+                            if (!tmpDevice.serialNumber.equals(serialNumber)) {
+                                LinkedBlockingQueue<QueueObject> tmpQueueObjects = devices.get(tmpDevice.serialNumber);
+                                QueueObject tmpQueueObject = tmpQueueObjects.peek();
+                                if (!tmpQueueObject.equals(queueObject)
+                                        || (tmpQueueObject.future != null && !queueObject.future.isDone())) {
+                                    execute = false;
+                                    break;
+                                }
+                                serial = serial + tmpDevice.serialNumber + " ";
+                            }
+                        }
+                        if (execute) {
+                            queueObject.future = scheduler.submit(() -> queuedExecuteSequenceNode(queueObject));
+                            logger.debug("thread {} device {}", queueObject.hashCode(), serial);
+                        }
+                    }
```

Comment:
Could this be implemented using CompletableFuture instead?

Reply:
Probably. But I would prefer to leave it as-is now as this part has a long history of wrong attempts to fix problems. It would be much easier to review the changes and see if they are really equivalent if there is nothing else in the PR.